nemo/collections/tts/modules/submodules.py

```diff
@@ -410,3 +419,284 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
+        weight = torch.nn.functional.softmax(weight, dim=-1)
+        speaker_emb = weight @ self.pretrained_embedding
+        return speaker_emb
+
+
+class GlobalStyleToken(NeuralModule):
+    """
+    Global Style Token based Speaker Embedding
+    """
+
+    def __init__(
+        self,
+        cnn_filters=[32, 32, 64, 64, 128, 128],
+        dropout=0.2,
+        gru_hidden=128,
+        gst_size=128,
+        n_style_token=10,
+        n_style_attn_head=4,
+    ):
+        super(GlobalStyleToken, self).__init__()
+        self.reference_encoder = ReferenceEncoder(
+            cnn_filters=list(cnn_filters), dropout=dropout, gru_hidden=gru_hidden
+        )
+        self.style_attention = StyleAttention(
+            gru_hidden=gru_hidden, gst_size=gst_size, n_style_token=n_style_token, n_style_attn_head=n_style_attn_head
+        )
+
+    @property
+    def input_types(self):
+        return {
+            "inp": NeuralType(('B', 'D', 'T_spec'), MelSpectrogramType()),
+            "inp_mask": NeuralType(('B', 'T_spec', 1), TokenDurationType()),
+        }
+
+    @property
+    def output_types(self):
+        return {
+            "gst": NeuralType(('B', 'D'), EncodedRepresentation()),
+        }
+
+    def forward(self, inp, inp_mask):
+        style_embedding = self.reference_encoder(inp, inp_mask)
+        gst = self.style_attention(style_embedding)
+        return gst
+
+
+class ReferenceEncoder(NeuralModule):
+    """
+    Encode mel-spectrograms to an utterance level feature
+    """
+
+    def __init__(self, cnn_filters=[32, 32, 64, 64, 128, 128], dropout=0.2, gru_hidden=128):
+        super(ReferenceEncoder, self).__init__()
+        self.filter_size = [1] + cnn_filters
+        self.dropout = dropout
+        self.conv = torch.nn.Sequential(
+            OrderedDict(
+                [
+                    module
+                    for i in range(len(cnn_filters))
+                    for module in (
+                        ("transpose_bchw_{}".format(i + 1), Transpose([0, 3, 1, 2])),
+                        (
+                            "conv2d_{}".format(i + 1),
+                            torch.nn.Conv2d(
+                                in_channels=int(self.filter_size[i]),
+                                out_channels=int(self.filter_size[i + 1]),
+                                kernel_size=(3, 3),
+                                stride=(2, 2),
+                                padding=(1, 1),
+                                dilation=(1, 1),
+                                bias=True,
+                            ),
+                        ),
+                        ("transpose_bhwc_{}".format(i + 1), Transpose([0, 2, 3, 1])),
+                        ("relu_{}".format(i + 1), torch.nn.ReLU()),
+                        ("layer_norm_{}".format(i + 1), torch.nn.LayerNorm(self.filter_size[i + 1]),),
+                        ("dropout_{}".format(i + 1), torch.nn.Dropout(self.dropout)),
```

Comment:
Would be convenient here and in the future if we have a generic Conv2dNorm layer that handles the awkward transposes and makes the activation/layer_norm optional.

Reply:
Binary dropout is basically designed to be used on relu activation output (effectively setting active neurons to inactive). Applying it after layer norm rather than after relu is strange.

Reply:
@rlangman  For this, do you know why we add dropout after norm?
https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/modules/fastpitch.py#L92